src/VisualStudio/Core/Def/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs

```diff
@@ -394,15 +405,15 @@ private async Task UpdateContextAfterOpenAsync(string filePath)
                 EnsureSuggestedActionsSourceProviderEnabled();
             }
 
-            internal void CheckForOpenFilesThatWeMissed()
+            internal async Task CheckForOpenFilesThatWeMissedAsync(CancellationToken cancellationToken)
             {
                 // It's possible that Roslyn is loading asynchronously after documents were already opened by the user; this is a one-time check for
                 // any of those -- after this point, we are subscribed to events so we'll know of anything else.
-                _foregroundAffinitization.AssertIsForeground();
+                await _foregroundAffinitization.ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
 
                 foreach (var (filePath, textBuffer, hierarchy) in _openTextBufferProvider.EnumerateDocumentSet())
                 {
-                    TryOpeningDocumentsForMonikerAndSetContextOnUIThread(filePath, textBuffer, hierarchy);
```

Comment:
📝 This call attempted to take a lock that was held by another thread, and blocked the UI

Reply:
@sharwell I'm a bit confused here -- we're still calling the same code more or less which is acquiring the same locks. So I'd imagine the deadlock you saw would still be able to happen. Or by "deadlock" did you just mean we're potentially blocking the UI thread synchronously when that wasn't necessary but we'd eventually make progress?

Reply:
Instead of blocking the UI thread while waiting to acquire the workspace lock, we now block a background thread.